src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs

```diff
@@ -91,24 +92,69 @@ public static CodeWriter WriteFieldDeclaration(this CodeWriter writer, FieldDecl
 
         public static CodeWriter.CodeWriterScope WriteMethodDeclaration(this CodeWriter writer, MethodSignatureBase methodBase, params string[] disabledWarnings)
         {
-            foreach (var disabledWarning in disabledWarnings)
+            WriteDisableWarnings(writer, disabledWarnings);
+
+            writer.Append($"{methodBase.Modifiers} ");
+            if (methodBase is MethodSignature method)
             {
-                writer.Line($"#pragma warning disable {disabledWarning}");
+                if (method.ReturnType != null)
+                {
+                    writer.Append($"{method.ReturnType} ");
+                }
+                else
+                {
+                    writer.AppendRaw("void ");
+                }
             }
 
+            return WriteMethodDeclarationParameters(writer, methodBase, disabledWarnings, methodBase.Name);
+        }
+
+        public static CodeWriter.CodeWriterScope WriteMethodDeclaration(this CodeWriter writer, MethodSignatureBase methodBase, bool isAsync, params string[] disabledWarnings)
+        {
+            WriteDisableWarnings(writer, disabledWarnings);
+
             writer.Append($"{methodBase.Modifiers} ");
             if (methodBase is MethodSignature method)
             {
+                if (isAsync && !method.IsPageable)
+                    writer.Append($"async ");
+
+                var firstParam = method.Parameters.FirstOrDefault();
+                bool isExtensionMethod = firstParam is not null && firstParam.IsExtensionParameter;
+
+                if (method.Modifiers.Contains("public") && !isExtensionMethod)
+                    writer.Append($"virtual ");
+
+                if (isExtensionMethod)
+                    writer.Append($"static ");
+
                 if (method.ReturnType != null)
                 {
-                    writer.Append($"{method.ReturnType} ");
+                    var finalType = method.IsPageable ? method.ReturnType.WrapPageable(isAsync) : method.ReturnType.WrapAsync(isAsync);
+                    writer.Append($"{finalType} ");
                 }
                 else
                 {
                     writer.AppendRaw("void ");
                 }
             }
-            writer.Append($"{methodBase.Name}(");
+
+            string methodName = isAsync ? $"{methodBase.Name}Async" : methodBase.Name;
+            return WriteMethodDeclarationParameters(writer, methodBase, disabledWarnings, methodName);
+        }
+
+        private static void WriteDisableWarnings(CodeWriter writer, string[] disabledWarnings)
+        {
+            foreach (var disabledWarning in disabledWarnings)
+            {
+                writer.Line($"#pragma warning disable {disabledWarning}");
```

Comment:
Are these reenabled?  I just don't see where - expected it to appear in this method.

Reply:
https://github.com/Azure/autorest.csharp/pull/1940/files/4746d265e49fa51066ed62469935f8280628357f#diff-75cfcbfd3603fecfefa335f34da14c86ccd5d3a966d18937b64e08228ad9aaa0R185